### DISCUSSION TITLE
Possible glitch: Track titles containing /

### DISCUSSION THREAD

**Discussion by kubycsolutions** [automatic-ripping-machine/automatic-ripping-machine#1149] [Q&A]
Ripping some trad-flavored stuff, some tracks are medleys of multiple tunes. After ripping, this winds up as additional directories, some of them with spaces in their names. Example: A track named "Wandering Willie / The Wily Old Bachelor / Within a Mile of Clonbur" comes through as

```
arm@medea:~$ ls -dl cds/Heather\ Alexander/A\ Gypsy\'s\ Home\ -\ Songs\ of\ the\ Roving\ Life/06*
drwxr-xr-x 3 arm arm 4096 Jun 11 00:56 "cds/Heather Alexander/A Gypsy's Home - Songs of the Roving Life/06 - Wandering Willie "
arm@medea:~$ ls -l cds/Heather\ Alexander/A\ Gypsy\'s\ Home\ -\ Songs\ of\ the\ Roving\ Life/06*
total 4
drwxr-xr-x 2 arm arm 4096 Jun 11 00:56 ' The Wily Old Bachelor '
arm@medea:~$ ls -l cds/Heather\ Alexander/A\ Gypsy\'s\ Home\ -\ Songs\ of\ the\ Roving\ Life/06\ -\ Wandering\ Willie\ /
total 4
drwxr-xr-x 2 arm arm 4096 Jun 11 00:56 ' The Wily Old Bachelor '
arm@medea:~$ ls -l cds/Heather\ Alexander/A\ Gypsy\'s\ Home\ -\ Songs\ of\ the\ Roving\ Life/06\ -\ Wandering\ Willie\ /\ The\ Wily\ Old\ Bachelor\ /
total 26856
-rw-r--r-- 1 arm arm 27498661 Jun 11 00:56 ' Within a Mile of Clonbur.flac'
arm@medea:~$ 
```

It could be much worse, but maybe it would be better if we recognized that on audio album track titles, slash generally does not mean subdirectory?

**Comment by jkesselm** [comment]
Might be as simple as escaping `/` in track name (and album name?) when building the path. If I find cycles, I may take a look at the source and do a quick experiment...

This does raise the question of whether there are other things we should escape for safety.  Hopefully everything is going through APIs rather than ever getting near the shell interpreter, so `;` and `$` and `#` and so on should be harmless...?

**Reply by studog-github** [reply]
I was just looking at this because my `AC/DC - Who Made Who` cd ripped to `AC/DC/Who Made Who` (3 directories) instead of `AC_DC/Who Made Who` or something similar (2 directories).  
  
The issue appears to be with the `abcde` config options; they suffer from a "path separator" injection bug looks like. This isn't important to me right now, but may be in the future.  
  
I think the likely solution is having `abcde` updated to handle [ looked up data contains filesystem-illegal characters or reserved names ] by doing some escaping or replacement. "AC_DC/Who Made Who" would have been acceptable to me. I imagine this already exists in abcde but I haven't checked.  
  
There's a further larger issue: if you expect your ripped audio to live on various filesystems (rip on Linux, transfer to a FAT32 external drive, for example) what is really needed is the lowest common filesystem denominator. And add in a "supports UTF-8 or not" factor too. StackOverflow has a lot of discussion around this.

**Comment by JollyRgrs** [comment]
towards the bottom of the [abcde.conf](https://github.com/automatic-ripping-machine/automatic-ripping-machine/blob/ac74406e4b6c7d0da9936fa8563d76ddb1d0551e/setup/.abcde.conf#L475) file, there is an option for filename munging. I prefer starting with VARIANT 3 and tweaking how I want it. This does require a bit of knowledge of sed/regex magic (some great internet resources are available to help understand that some) but the default on variant 3 takes care of forward slashes for sure. I've had track/album/artists with `*`, `/`, and other various issues. I found a particular problem is when viewing my files on a SMB share on Windows, folders (or most likely files as well) with a trailing `.` cause issues. I've adjusted my filemunge() function to handle some of those cases.

You can tell it what to replace the characters with (or to just remove and replace with nothing).

**Reply by microtechno9000** [reply]
Agree, this is an abcde problem and the munge function fixes it.

@JollyRgrs can you post the function you are using, or raise a PR against the wiki with the information to share with others

**Reply by JollyRgrs** [reply]
> Agree, this is an abcde problem and the munge function fixes it.
> 
> @JollyRgrs can you post the function you are using, or raise a PR against the wiki with the information to share with others

sure thing. My current (this is always a work in progress)

```
mungefilename ()
{
        echo "$@" | sed 's/:/ - /g' | \
                        sed 's/\*/-/g' | \
                        sed 's/"//g' | \
                        sed 's/[\/@#\$%\^]/_/g' | \
                        sed 's/[\?!]//g' | \
                        sed 's/__\+/_/g' | \
                        sed 's/  \+/ /g' | \
                        sed 's/^_*//' | \
                        sed 's/_*$//'
}```
I am keeping both Linux and Windows in mind for these file names. Debating to only remove $ if it is at the end of the string (makes folder hidden in Windows). again, this is a WIP, but it should be a good start. using something like https://regex101.com/ can help you understand what the different things do. Note the default regex "flavor" doesn't work 100% in bash, so test things out first.

**Reply by studog-github** [reply]
Agree this is an abcde problem. Since ARM uses abcde, it is also an ARM problem.

The solution is to teach abcde about filesystem character and naming issues.

**Reply by JollyRgrs** [reply]
> Agree this is an abcde problem. Since ARM uses abcde, it is also an ARM problem.
> 
> The solution is to teach abcde about filesystem character and naming issues.

I agree and disagree. When you make a tool that uses ssh, you can't expect that tool to maintain any issues with ssh. However, as this is just a config issue, there are some options like possibly tweaking the default config to contain a more "sane" (opinions may vary) option for the munge function as a service for those using ARM. However, I also think it is important for ARM users to understand how this works so they can tweak it to their liking. I've, since posting above function, still tweaked mine further to remove trailing dots (Windows doesn't like folders ending in a "."). How to handle things like double dashes or what to replace a dollar sign or a forward slash with might be taken from some standards or (in my case) might just be personal preferences. This is why the option exists in the abcde.conf to begin with and I'm grateful for that. Perhaps we could come up with a list of all substitutions that we want to be default and see if it wouldn't be too much to ask to edit the config file for abcde.conf with those changes. Maybe a line about it in the config UI as well to make it easier to change, or at least pointing advanced users how to  edit the conf file to make any adjustments.

**Reply by studog-github** [reply]
Edit: I was typing this up while @JollyRgrs replied. My aim here is to help ARM users understand what the filemunging is doing, so they are able to alter the function to fit there preferences.

> sure thing. My current (this is always a work in progress)
> 
> ```
> mungefilename ()
> {
>         echo "$@" | sed 's/:/ - /g' | \             # Change colon to space-dash-space everywhere
>                         sed 's/\*/-/g' | \          # Change asterisk to dash everywhere
>                         sed 's/"//g' | \            # Change doublequote to nothing everywhere (effectively deleting)
>                         sed 's/[\/@#\$%\^]/_/g' | \ # Change (forwardslash, atsign, poundsign, dollarsign, percentsign, caret) to underscore everywhere
>                         sed 's/[\?!]//g' | \        # Change (questionmark, exclamationmark) to nothing everywhere (effectively deleting)
>                         sed 's/__\+/_/g' | \        # Change (sequence of two or more underscores in a row) to underscore everywhere
>                         sed 's/  \+/ /g' | \        # Change (sequence of two or more spaces in a row) to space everywhere
>                         sed 's/^_*//' | \           # Change (sequence of zero or more underscores at the start of the string) to nothing (effectively deleting)
>                         sed 's/_*$//'               # Change (sequence of zero or more underscores at the end of the string) to nothing (effectively deleting)
> }```

Added explanations.

> using something like https://regex101.com/ can help you understand what the different things do. Note the default regex "flavor" doesn't work 100% in bash, so test things out first.

I assume you mean _regex101.com's default flavour of regex_ there, as there is no universal default flavour of regex.

Bash has its own regex flavour but that isn't in use here.
sed has two, regular and extended. This function is using regular.

regex101 only has a few flavours available and is focused on web-programming flavours, so there's not a great match.  
  
The only good way to figure out what regex you need is to test it in the target flavour. This can be tedious though.  
  
Note that the above function is a bash/shell script, so [ quoting to protect against shell interpretation ] is another layer of complexity involved.

https://gist.github.com/CMCDragonkai/6c933f4a7d713ef712145c5eb94a1816 gives an idea of how messy regex is, in the wild.
